quark.c

```diff
@@ -4736,6 +4738,217 @@ raw_event_tty(struct quark_queue *qq, struct raw_event *raw)
 	return (qev);
 }
 
+void
+quark_ruleset_init(struct quark_ruleset *ruleset)
+{
+	ruleset->rules = NULL;
+	ruleset->n_rules = 0;
+}
+
+void
+quark_ruleset_clear(struct quark_ruleset *ruleset)
+{
+	struct quark_rule	*rule;
+	size_t			 i;
+
+	for (i = 0; i < ruleset->n_rules; i++) {
+		rule = ruleset->rules + i;
+		free(rule->fields);
+	}
+
+	free(ruleset->rules);
+	ruleset->rules = NULL;
+	ruleset->n_rules = 0;
+}
+
+static int
+path_match(struct rule_field *field, const char *a)
+{
+	if (field->wildcard_len == 0)
+		return (!strcmp(field->path, a));
+	else
+		return (!strncmp(field->path, a, field->wildcard_len));
+}
+
+/* 1 for match, 0 for non match */
+static int
+rule_field_match(struct quark_rule *rule, struct rule_field *field,
+    struct quark_event *qev)
+{
+	switch (field->code) {
+	case RF_PROCESS_PID:
+		if (qev->process != NULL)
+			return (qev->process->pid == field->pid);
+		break;
+	case RF_PROCESS_PPID:
+		if (qev->process != NULL)
+			return (qev->process->proc_ppid == field->pid);
+		break;
+	case RF_PROCESS_FILENAME:
+		if (qev->process != NULL &&
+		    (qev->process->flags & QUARK_F_FILENAME) &&
+		    qev->process->filename != NULL)
+			return (path_match(field, qev->process->filename));
+		break;
+	case RF_FILE_PATH:
+		if (qev->file != NULL)
```

Comment:
I notice that you checked `qev->process->filename != NULL`, why not `qev->file->path != NULL` here?

Reply:
basically file->path cannot be null, it points to itself